add brief results/prjresults with filtered code

lethliel opened the pull request:
in do_results:
  * add --brief option on prj level:
    [packagename] [repo] [arch] [buildstatus]
  * filter by --status-filter <long status name>
    works on prj and pkg level

in do_prjresults:
  * --brief
  * assume len(state)>1 as long state

core.py
  * filter packages by build status
  * long status handling in get_prj_results
  * brief output generation in get_prj_results

implements https://github.com/openSUSE/osc/issues/581

marcus-h commented:
On 2020-06-18 23:44:23 -0700, Marco Strigl wrote:
> in do_results:
>   * add --brief option on prj level:
>     [packagename] [repo] [arch] [buildstatus]
>   * filter by --status <long status name>
>     works on prj and pkg level
>
> in do_prjresults:
>   * --brief and --status
>   * assume len(state)>1 as long state
>
Hmm just a thought: if each change goes into a separate commit, it would make
the review a bit easier:)

Below, just a very quick high-level review.

> diff --git a/osc/commandline.py b/osc/commandline.py
> index a172d850..945261d7 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -5259,6 +5259,8 @@ def do_rremove(self, subcmd, opts, project, package, *files):
>                          help='Show results only for specified repo(s)')
>      @cmdln.option('-a', '--arch', action='append', default = [],
>                          help='Show results only for specified architecture(s)')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,
> +                        help='show the result in "pkgename repo arch result". Default for -f')

Typo? pkgename -> pkgname?

>      @cmdln.option('-v', '--verbose', action='store_true', default=False,
>                          help='more verbose output')
>      @cmdln.option('--no-multibuild', action='store_true', default=False,
> @@ -5269,6 +5271,10 @@ def do_rremove(self, subcmd, opts, project, package, *files):
>                          help='list packages vertically instead horizontally for entire project')
>      @cmdln.option('-w', '--watch', action='store_true',
>                          help='watch the results until all finished building')
> +    @cmdln.option('-s', '--status', default = None,
> +                        help='watch the results until all finished building')

Wrong help text.

> +    @cmdln.option('-f', '--failed', action='store_true',
> +                        help='watch the results until all finished building')

Wrong help text.

>      @cmdln.option('', '--xml', action='store_true', default=False,
>                          help='generate output in XML (former results_meta)')
>      @cmdln.option('', '--csv', action='store_true', default=False,
> @@ -5307,6 +5313,13 @@ def do_results(self, subcmd, opts, *args):
>          if project == None:
>              raise oscerr.WrongOptions("No project given")
>
> +        if opts.failed and opts.status:
> +            raise oscerr.WrongArgs('-s and -f cannot be used together')
> +
> +        if opts.failed:
> +            opts.status = 'failed'
> +            opts.brief = True
> +
>          if package == None:
>              opts.hide_legend = None
>              opts.name_filter = None
> @@ -5320,7 +5333,8 @@ def do_results(self, subcmd, opts, *args):
>
>          kwargs = {'apiurl': apiurl, 'project': project, 'package': package,
>                    'lastbuild': opts.last_build, 'repository': opts.repo,
> -                  'arch': opts.arch, 'wait': opts.watch, 'showexcl': opts.show_excluded}
> +                  'arch': opts.arch, 'wait': opts.watch, 'showexcl': opts.show_excluded,
> +                  'status': opts.status}
>          if opts.multibuild_package:
>              opts.no_multibuild = False
>              kwargs['multibuild_packages'] = opts.multibuild_package
> @@ -5348,6 +5362,8 @@ def do_results(self, subcmd, opts, *args):
>      #          as well when adding a new option!
>      @cmdln.option('-q', '--hide-legend', action='store_true',
>                          help='hide the legend')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,
> +                        help='show the result in "pkgename repo arch result". Default for -f')

Typo? pkgename -> pkgname? (See above)

>      @cmdln.option('-w', '--watch', action='store_true',
>                          help='watch the results until all finished building, only supported with --xml')
>      @cmdln.option('-c', '--csv', action='store_true',
> @@ -5366,6 +5382,8 @@ def do_results(self, subcmd, opts, *args):
>                          help='list packages vertically instead horizontally')
>      @cmdln.option('--show-excluded', action='store_true',
>                          help='show packages that are excluded in all repos, also hide repos that have only excluded packages')
> +    @cmdln.option('--status', default=None,
> +                        help='build status of the package. (long version "succeeded" for example')

Hmm why don't we simply use the existing --status-filter?

>      @cmdln.alias('pr')
>      def do_prjresults(self, subcmd, opts, *args):
>          """${cmd_name}: Shows project-wide build results
> @@ -5387,6 +5405,10 @@ def do_prjresults(self, subcmd, opts, *args):
>              wd = os.curdir
>              project = store_read_project(wd)
>
> +        if opts.status_filter and len(opts.status_filter) > 1:
> +            opts.status = opts.status_filter
> +            opts.status_filter = None
> +
>          if opts.xml:
>              kwargs = {}
>              if opts.repo:
> @@ -5406,7 +5428,8 @@ def do_prjresults(self, subcmd, opts, *args):
>                                          csv=opts.csv, status_filter=opts.status_filter, \
>                                          name_filter=opts.name_filter, repo=opts.repo, \
>                                          arch=opts.arch, vertical=opts.vertical, \
> -                                        show_excluded=opts.show_excluded)))
> +                                        show_excluded=opts.show_excluded, brief=opts.brief, \
> +                                        lstatus=opts.status)))
>
>      @cmdln.option('-q', '--hide-legend', action='store_true',
>                          help='hide the legend')
> diff --git a/osc/core.py b/osc/core.py
> index 9d032a1e..db7aecb1 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -5695,7 +5695,7 @@ def get_binarylist_published(apiurl, prj, repo, arch):
>      return r
>
>
> -def show_results_meta(apiurl, prj, package=None, lastbuild=None, repository=[], arch=[], oldstate=None, multibuild=False, locallink=False):
> +def show_results_meta(apiurl, prj, package=None, lastbuild=None, repository=[], arch=[], oldstate=None, multibuild=False, locallink=False, status=None):

I think we should go with "code" instead of "status" here.

>      query = []
>      if package:
>          query.append('package=%s' % quote_plus(package))
> @@ -5707,6 +5707,8 @@ def show_results_meta(apiurl, prj, package=None, lastbuild=None, repository=[],
>          query.append('multibuild=1')
>      if locallink:
>          query.append('locallink=1')
> +    if status:
> +        query.append('code=%s' % quote_plus(status))
>      for repo in repository:
>          query.append('repository=%s' % quote_plus(repo))
>      for a in arch:
> @@ -5811,11 +5813,11 @@ def get_results(apiurl, project, package, verbose=False, printJoin='', *args, **
>                      res['status'] += '(unpublished)'
>                  else:
>                      res['status'] += '*'
> -
> -            if is_multi:
> -                r.append(result_line_mb_templ % res)
> -            else:
> -                r.append(result_line_templ % res)
> +            if kwargs['status'] == res['status'] or not kwargs['status']:
> +                if is_multi:
> +                    r.append(result_line_mb_templ % res)
> +                else:
> +                    r.append(result_line_templ % res)
>
Why do we filter again? Didn't the backend to the filtering for us?
Also, it should be "res['code']" instead of "res['status']" (the "status"
key may not correspond to the "code" anymore).


>          if printJoin:
>              if printed:
> @@ -5868,7 +5870,7 @@ def get_package_results(apiurl, project, package=None, wait=False, *args, **kwar
>      yield xml
>
>
> -def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None):
> +def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None, brief=False, lstatus=None):
>      #print '----------------------------------------'
>      global buildstatus_symbols
>
> @@ -5909,13 +5911,12 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>
>      # filter option
>      if status_filter or name_filter or not show_excluded:
> -
>          pacs_to_show = []
>          targets_to_show = []
>
>          #filtering for Package Status
> -        if status_filter:
> -            if status_filter in buildstatus_symbols.values():
> +        if status_filter or lstatus:
> +            if status_filter in buildstatus_symbols.values() or lstatus:
>                  # a list is needed because if status_filter == "U"
>                  # we have to filter either an "expansion error" (obsolete)
>                  # or an "unresolvable" state
> @@ -5923,6 +5924,8 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>                  for txt, sym in buildstatus_symbols.items():
>                      if sym == status_filter:
>                          filters.append(txt)
> +                    elif lstatus:
> +                        filters.append(lstatus)

Hmm there's no need to this in a loop (see below).

>                  for filt_txt in filters:
>                      for pkg in status.keys():
>                          for repo in status[pkg].keys():
> @@ -5968,6 +5971,20 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>              r.append(';'.join(row))
>          return r
>
> +    if brief:
> +        for pac in pacs:
> +            for tg in targets:
> +                if tg in status[pac]:
> +                    if status_filter or lstatus:
> +                        for fill in filters:
> +                            if fill == status[pac][tg]:
> +                                row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
> +

Doesn't the same entry show up several times now? (At least, lstatus
exists in the "filters" list several times.)

> +                    else:
> +                        row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
> +                    r.append(' '.join(row))
> +        return r
> +
>      if not vertical:
>          # human readable output
>          max_pacs = 40
>

lethliel commented:
@marcus-h Changes as discussed via IRC done.

marcus-h commented:
On 2020-06-26 02:14:29 -0700, Marco Strigl wrote:
> diff --git a/osc/commandline.py b/osc/commandline.py
> index a172d850..e6c7960c 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -5259,6 +5259,8 @@ def do_rremove(self, subcmd, opts, project, package, *files):
>                          help='Show results only for specified repo(s)')
>      @cmdln.option('-a', '--arch', action='append', default = [],
>                          help='Show results only for specified architecture(s)')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,
> +                        help='show the result in "pkgname repo arch result". Default for -f')

Out of curiosity: is there a specific reason why this is the default for "-f"?

...

> @@ -5348,6 +5361,8 @@ def do_results(self, subcmd, opts, *args):
>      #          as well when adding a new option!
>      @cmdln.option('-q', '--hide-legend', action='store_true',
>                          help='hide the legend')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,
> +                        help='show the result in "pkgname repo arch result". Default for -f')

The "Default for..." part should be removed.

>      @cmdln.option('-w', '--watch', action='store_true',
>                          help='watch the results until all finished building, only supported with --xml')
>      @cmdln.option('-c', '--csv', action='store_true',
> @@ -5387,6 +5402,11 @@ def do_prjresults(self, subcmd, opts, *args):
>              wd = os.curdir
>              project = store_read_project(wd)
>
> +        lstatus = None
> +        if opts.status_filter and len(opts.status_filter) > 1:
> +            lstatus = opts.status_filter
> +            opts.status_filter = None
> +

Let's get rid of lstatus (see below).

...

> diff --git a/osc/core.py b/osc/core.py
> index 9d032a1e..6fcb4d7e 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -5811,11 +5813,11 @@ def get_results(apiurl, project, package, verbose=False, printJoin='', *args, **
>                      res['status'] += '(unpublished)'
>                  else:
>                      res['status'] += '*'
> -
> -            if is_multi:
> -                r.append(result_line_mb_templ % res)
> -            else:
> -                r.append(result_line_templ % res)
> +            if kwargs['code'] == res['code'] or not kwargs['code']:

This breaks if kwargs has no "code" key. Before the loop you could do
something like

code_filter = kwargs.get('code')

and then in the loop

if code_filter is None or code_filter == res['code']: ...

Can you also add a small comment we need to do the filtering? (The
result_xml_to_dicts stuff that you told me on IRC.)

> +                if is_multi:
> +                    r.append(result_line_mb_templ % res)
> +                else:
> +                    r.append(result_line_templ % res)
>
>          if printJoin:
>              if printed:
> @@ -5868,7 +5870,7 @@ def get_package_results(apiurl, project, package=None, wait=False, *args, **kwar
>      yield xml
>
>
> -def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None):
> +def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None, brief=False, lstatus=None):
>      #print '----------------------------------------'
>      global buildstatus_symbols
>
> @@ -5909,20 +5911,22 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>
>      # filter option
>      if status_filter or name_filter or not show_excluded:
> -
>          pacs_to_show = []
>          targets_to_show = []
>
>          #filtering for Package Status
> -        if status_filter:
> -            if status_filter in buildstatus_symbols.values():
> +        if status_filter or lstatus:
> +            if status_filter in buildstatus_symbols.values() or lstatus:
>                  # a list is needed because if status_filter == "U"
>                  # we have to filter either an "expansion error" (obsolete)
>                  # or an "unresolvable" state
>                  filters = []
> -                for txt, sym in buildstatus_symbols.items():
> -                    if sym == status_filter:
> -                        filters.append(txt)
> +                if lstatus:
> +                    filters.append(lstatus)
> +                else:
> +                    for txt, sym in buildstatus_symbols.items():
> +                        if sym == status_filter:
> +                            filters.append(txt)

We do not need the lstatus parameter. We can simply pass the "long" status
via the status_filter, too.

if status_filter:
    filters = []
    if status_filter in buildstatus_symbols.values():
        # old "value to key mapping" code here
    else:
        filters.append(status_filter)
    # filtering code...

The question is if we still need to keep the status_filter code path.
Actually, the backend could do the code filtering. We just need to keep
the name_filter stuff.
(I'm also fine if you want to keep the code as is - we could do the
refactoring later.)

>                  for filt_txt in filters:
>                      for pkg in status.keys():
>                          for repo in status[pkg].keys():
> @@ -5932,7 +5936,6 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>                                      targets_to_show.append(repo)
>                                  elif name_filter in pkg:
>                                      pacs_to_show.append(pkg)
> -
>          #filtering for Package Name
>          elif name_filter:
>              for pkg in pacs:
> @@ -5968,6 +5971,24 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>              r.append(';'.join(row))
>          return r
>
> +    if brief:
> +        for pac in pacs:
> +            for tg in targets:
> +                outer_loop_continue = False
> +                if status_filter or lstatus:
> +                    for fill in filters:
> +                        if tg in status[pac] and fill == status[pac][tg]:
> +                            row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
> +                        else:
> +                            outer_loop_continue = True
> +                            break
> +                    if outer_loop_continue == True:
> +                        continue
> +                else:
> +                    row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
> +                r.append(' '.join(row))

There's no need to do the filtering again. Each element of pacs already
survived the (status/code) filtering. (Also, there is no need to do
this list concatenation stuff:) )

lethliel commented:
On 6/30/20 1:49 PM, Marcus Hüwe wrote:
> On 2020-06-26 02:14:29 -0700, Marco Strigl wrote:
>> diff --git a/osc/commandline.py b/osc/commandline.py
>> index a172d850..e6c7960c 100644
>> --- a/osc/commandline.py
>> +++ b/osc/commandline.py
>> @@ -5259,6 +5259,8 @@ def do_rremove(self, subcmd, opts, project,
> package, *files):
>> help='Show results only for specified repo(s)')
>> @cmdln.option('-a', '--arch', action='append', default = [],
>> help='Show results only for specified architecture(s)')
>> + @cmdln.option('-b', '--brief', action='store_true', default=False,
>> + help='show the result in "pkgname repo arch result". Default for -f')
> 
> Out of curiosity: is there a specific reason why this is the default for
> "-f"?

No. Just thought that would be the expected behavior of the user. So
this is basically an assumption from me.

> 
> ...
> 
>> @@ -5348,6 +5361,8 @@ def do_results(self, subcmd, opts, *args):
>> # as well when adding a new option!
>> @cmdln.option('-q', '--hide-legend', action='store_true',
>> help='hide the legend')
>> + @cmdln.option('-b', '--brief', action='store_true', default=False,
>> + help='show the result in "pkgname repo arch result". Default for -f')
> 
> The "Default for..." part should be removed.

removed.

> 
>> @cmdln.option('-w', '--watch', action='store_true',
>> help='watch the results until all finished building, only supported
> with --xml')
>> @cmdln.option('-c', '--csv', action='store_true',
>> @@ -5387,6 +5402,11 @@ def do_prjresults(self, subcmd, opts, *args):
>> wd = os.curdir
>> project = store_read_project(wd)
>>
>> + lstatus = None
>> + if opts.status_filter and len(opts.status_filter) > 1:
>> + lstatus = opts.status_filter
>> + opts.status_filter = None
>> +
> 
> Let's get rid of lstatus (see below).
> 
> ...
> 
>> diff --git a/osc/core.py b/osc/core.py
>> index 9d032a1e..6fcb4d7e 100644
>> --- a/osc/core.py
>> +++ b/osc/core.py
>> @@ -5811,11 +5813,11 @@ def get_results(apiurl, project, package,
> verbose=False, printJoin='', *args, **
>> res['status'] += '(unpublished)'
>> else:
>> res['status'] += '*'
>> -
>> - if is_multi:
>> - r.append(result_line_mb_templ % res)
>> - else:
>> - r.append(result_line_templ % res)
>> + if kwargs['code'] == res['code'] or not kwargs['code']:
> 
> This breaks if kwargs has no "code" key. Before the loop you could do
> something like
> 
> code_filter = kwargs.get('code')
> 
> and then in the loop
> 
> if code_filter is None or code_filter == res['code']: ...
> 
> Can you also add a small comment we need to do the filtering? (The
> result_xml_to_dicts stuff that you told me on IRC.)

done and fixed the case when kwargs['code'] does not exist.

> 
>> + if is_multi:
>> + r.append(result_line_mb_templ % res)
>> + else:
>> + r.append(result_line_templ % res)
>>
>> if printJoin:
>> if printed:
>> @@ -5868,7 +5870,7 @@ def get_package_results(apiurl, project,
> package=None, wait=False, *args, **kwar
>> yield xml
>>
>>
>> -def get_prj_results(apiurl, prj, hide_legend=False, csv=False,
> status_filter=None, name_filter=None, arch=None, repo=None,
> vertical=None, show_excluded=None):
>> +def get_prj_results(apiurl, prj, hide_legend=False, csv=False,
> status_filter=None, name_filter=None, arch=None, repo=None,
> vertical=None, show_excluded=None, brief=False, lstatus=None):
>> #print '----------------------------------------'
>> global buildstatus_symbols
>>
>> @@ -5909,20 +5911,22 @@ def get_prj_results(apiurl, prj,
> hide_legend=False, csv=False, status_filter=Non
>>
>> # filter option
>> if status_filter or name_filter or not show_excluded:
>> -
>> pacs_to_show = []
>> targets_to_show = []
>>
>> #filtering for Package Status
>> - if status_filter:
>> - if status_filter in buildstatus_symbols.values():
>> + if status_filter or lstatus:
>> + if status_filter in buildstatus_symbols.values() or lstatus:
>> # a list is needed because if status_filter == "U"
>> # we have to filter either an "expansion error" (obsolete)
>> # or an "unresolvable" state
>> filters = []
>> - for txt, sym in buildstatus_symbols.items():
>> - if sym == status_filter:
>> - filters.append(txt)
>> + if lstatus:
>> + filters.append(lstatus)
>> + else:
>> + for txt, sym in buildstatus_symbols.items():
>> + if sym == status_filter:
>> + filters.append(txt)
> 
> We do not need the lstatus parameter. We can simply pass the "long" status
> via the status_filter, too.
> 
> if status_filter:
> filters = []
> if status_filter in buildstatus_symbols.values():
> # old "value to key mapping" code here
> else:
> filters.append(status_filter)
> # filtering code...
> 
> The question is if we still need to keep the status_filter code path.
> Actually, the backend could do the code filtering. We just need to keep
> the name_filter stuff.
> (I'm also fine if you want to keep the code as is - we could do the
> refactoring later.)
> 

done. I would like to keep it as it is and do the refactoring later.

>> for filt_txt in filters:
>> for pkg in status.keys():
>> for repo in status[pkg].keys():
>> @@ -5932,7 +5936,6 @@ def get_prj_results(apiurl, prj,
> hide_legend=False, csv=False, status_filter=Non
>> targets_to_show.append(repo)
>> elif name_filter in pkg:
>> pacs_to_show.append(pkg)
>> -
>> #filtering for Package Name
>> elif name_filter:
>> for pkg in pacs:
>> @@ -5968,6 +5971,24 @@ def get_prj_results(apiurl, prj,
> hide_legend=False, csv=False, status_filter=Non
>> r.append(';'.join(row))
>> return r
>>
>> + if brief:
>> + for pac in pacs:
>> + for tg in targets:
>> + outer_loop_continue = False
>> + if status_filter or lstatus:
>> + for fill in filters:
>> + if tg in status[pac] and fill == status[pac][tg]:
>> + row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
>> + else:
>> + outer_loop_continue = True
>> + break
>> + if outer_loop_continue == True:
>> + continue
>> + else:
>> + row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
>> + r.append(' '.join(row))
> 
> There's no need to do the filtering again. Each element of pacs already
> survived the (status/code) filtering. (Also, there is no need to do
> this list concatenation stuff:) )
> 

Thought so too. But if I switch to:

for pac in pacs:
    for tg in targets:
        if tg in status[pac]:
	    row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
        r.append(' '.join(row))
return r

I also get the other states:

build openSUSE_Factory i586 failed
build openSUSE_Factory x86_64 failed
build openSUSE_Factory_ARM aarch64 failed
build openSUSE_Factory_ARM armv7l failed
build openSUSE_Leap_15.0 x86_64 succeeded
build openSUSE_Leap_15.1 x86_64 succeeded
build openSUSE_Leap_42.1 x86_64 succeeded
build openSUSE_Leap_42.2 x86_64 succeeded
build openSUSE_Leap_42.3 x86_64 succeeded
[...]

And if I readd the filtering to "as it is" it works correctly.

So If you have a clue what's going wrong here I am more than happy to
get rid of the for fill in filters: loop.

lethliel commented:
@marcus-h updated and commented.

marcus-h commented:
On 2020-07-02 01:42:21 -0700, Marco Strigl wrote:
> >> @@ -5968,6 +5971,24 @@ def get_prj_results(apiurl, prj,

...

> > There's no need to do the filtering again. Each element of pacs already
> > survived the (status/code) filtering. (Also, there is no need to do
> > this list concatenation stuff:) )
> >
>
> Thought so too. But if I switch to:
>
> for pac in pacs:
>     for tg in targets:
>         if tg in status[pac]:
> 	    row = [pac] + [' '.join(tg[:-1])] + [status[pac][tg]]
>         r.append(' '.join(row))
> return r
>
> I also get the other states:
>
> build openSUSE_Factory i586 failed
> build openSUSE_Factory x86_64 failed
> build openSUSE_Factory_ARM aarch64 failed
> build openSUSE_Factory_ARM armv7l failed
> build openSUSE_Leap_15.0 x86_64 succeeded
> build openSUSE_Leap_15.1 x86_64 succeeded
> build openSUSE_Leap_42.1 x86_64 succeeded
> build openSUSE_Leap_42.2 x86_64 succeeded
> build openSUSE_Leap_42.3 x86_64 succeeded
> [...]
>
> And if I readd the filtering to "as it is" it works correctly.
>
> So If you have a clue what's going wrong here I am more than happy to
> get rid of the for fill in filters: loop.
>
Arghs - you are completely right:) As soon as a repo has, for instance,
a package with code=failed, it will be shown for all packages (regardless
of their "code" within the repo). Sorry for the confusion.
What about the following?

if brief:
    for pac, repo_states in status.items():
        for repo, state in repo_states.items():
            if filters and state not in filters:
                continue
            r.append('%s %s %s %s' % (pac, repo[0], repo[1], state))
    return r

(it is probably the best if we move the "filters = []" definition ot ouf
the two if statements (move it after the "# filter option" comment) (maybe
we should rename it to "status_filters" (or "code_filters")))

Below two minor comments about the current version:

> @@ -5811,11 +5814,14 @@ def get_results(apiurl, project, package, verbose=False, printJoin='', *args, **
>                      res['status'] += '(unpublished)'
>                  else:
>                      res['status'] += '*'
> -
> -            if is_multi:
> -                r.append(result_line_mb_templ % res)
> -            else:
> -                r.append(result_line_templ % res)
> +            # we need to do the code filtering again, because result_xml_to_dicts returns the code
> +            # of the repository if the result is already prefiltered by the backend. So we need
> +            # to filter out the repository states.
> +            if code_filter is None or kwargs['code'] == res['code']:

Hmm just use code_filter instead of kwargs['code'].

>                  for txt, sym in buildstatus_symbols.items():
> -                    if sym == status_filter:
> -                        filters.append(txt)

...

> +                        if sym == status_filter:
> +                            filters.append(txt)

Over-indented.

marcus-h commented:
Great! There is only one small issue left (the other remarks are just
beautifications (feel free to ignore:) )).

On 2020-07-02 01:46:53 -0700, Marco Strigl wrote:
> diff --git a/osc/commandline.py b/osc/commandline.py
> index a172d850..405bae85 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -5259,6 +5259,8 @@ def do_rremove(self, subcmd, opts, project, package, *files):
>                          help='Show results only for specified repo(s)')
>      @cmdln.option('-a', '--arch', action='append', default = [],
>                          help='Show results only for specified architecture(s)')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,

The default=False can be omitted.

> +                        help='show the result in "pkgname repo arch result". Default for -f')
>      @cmdln.option('-v', '--verbose', action='store_true', default=False,
>                          help='more verbose output')
>      @cmdln.option('--no-multibuild', action='store_true', default=False,
> @@ -5269,6 +5271,10 @@ def do_rremove(self, subcmd, opts, project, package, *files):
>                          help='list packages vertically instead horizontally for entire project')
>      @cmdln.option('-w', '--watch', action='store_true',
>                          help='watch the results until all finished building')
> +    @cmdln.option('-s', '--status-filter', default = None,

The "default = None" can be omitted (there should be no whitespaces around "="
(I know... some options here still have the whitespaces:/)).

> +                        help='only show packages with the given build status')

...

> @@ -5348,6 +5361,8 @@ def do_results(self, subcmd, opts, *args):
>      #          as well when adding a new option!
>      @cmdln.option('-q', '--hide-legend', action='store_true',
>                          help='hide the legend')
> +    @cmdln.option('-b', '--brief', action='store_true', default=False,

default=False can be omitted (see above).

> +                        help='show the result in "pkgname repo arch result"')
>      @cmdln.option('-w', '--watch', action='store_true',
>                          help='watch the results until all finished building, only supported with --xml')
>      @cmdln.option('-c', '--csv', action='store_true',


> diff --git a/osc/core.py b/osc/core.py
> index 9d032a1e..8bce8337 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -5909,30 +5915,30 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>
>      # filter option
>      if status_filter or name_filter or not show_excluded:
> -
>          pacs_to_show = []
>          targets_to_show = []
>
>          #filtering for Package Status
>          if status_filter:
> +            filters = []

Can you move it one level up? That is,

filters = []
if status_filter or ...

(otherwise "osc prjresults -b ..." would break)

lethliel commented:
@marcus-h Could you please review again?

marcus-h commented:
On 2020-07-05 23:43:02 -0700, Marco Strigl wrote:
> diff --git a/osc/core.py b/osc/core.py
> index 9d032a1e..af654a39 100644
> --- a/osc/core.py
> +++ b/osc/core.py

...

> @@ -5868,7 +5874,7 @@ def get_package_results(apiurl, project, package=None, wait=False, *args, **kwar
>      yield xml
>
>
> -def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None):
> +def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=None, name_filter=None, arch=None, repo=None, vertical=None, show_excluded=None, brief=False):
>      #print '----------------------------------------'
>      global buildstatus_symbols
>
> @@ -5909,30 +5915,30 @@ def get_prj_results(apiurl, prj, hide_legend=False, csv=False, status_filter=Non
>
>      # filter option

filters = [] should be moved here (otherwise "osc prjresults -b ..." would
break).

>      if status_filter or name_filter or not show_excluded:
> -
>          pacs_to_show = []
>          targets_to_show = []
>
>          #filtering for Package Status
> +        filters = []

See above.

marcus-h commented:
Merged. Thanks a lot! (Also, thanks for your patience:) )
